src/builder/resources.rs

```diff
@@ -0,0 +1,218 @@
+            mem_request: self.mem_request,
+            other: self.other,
+            state: PhantomData,
+        }
+    }
+
+    pub fn with_memory_limit(
+        self,
+        limit: Quantity,
+    ) -> ResourceRequirementsBuilder<state::MissingCpuLimit> {
+        ResourceRequirementsBuilder {
+            cpu_limit: self.cpu_limit,
+            cpu_request: self.cpu_request,
+            mem_limit: Some(limit),
+            mem_request: self.mem_request,
+            other: self.other,
+            state: PhantomData,
+        }
+    }
+}
+
+impl ResourceRequirementsBuilder<state::MissingCpuLimit> {
+    pub fn with_cpu_limit(self, limit: Quantity) -> ResourceRequirementsBuilder<state::Final> {
+        ResourceRequirementsBuilder {
+            cpu_limit: Some(limit),
+            cpu_request: self.cpu_request,
+            mem_limit: self.mem_limit,
+            mem_request: self.mem_request,
+            other: self.other,
+            state: PhantomData,
+        }
+    }
+}
+
+impl ResourceRequirementsBuilder<state::MissingMemLimit> {
+    pub fn with_memory_limit(self, limit: Quantity) -> ResourceRequirementsBuilder<state::Final> {
+        ResourceRequirementsBuilder {
+            cpu_limit: self.cpu_limit,
+            cpu_request: self.cpu_request,
+            mem_limit: Some(limit),
+            mem_request: self.mem_request,
+            other: self.other,
+            state: PhantomData,
+        }
+    }
+}
+
+impl ResourceRequirementsBuilder<state::Final> {
+    pub fn build(self) -> ResourceRequirements {
+        let mut limits: BTreeMap<String, Quantity> = BTreeMap::new();
+        let mut requests: BTreeMap<String, Quantity> = BTreeMap::new();
+
+        // Insert the CPU and memory limits/requests only when they are set
+        if let Some(cpu_limit) = self.cpu_limit {
+            limits.insert("cpu".into(), cpu_limit);
+        }
+
+        if let Some(mem_limit) = self.mem_limit {
+            limits.insert("memory".into(), mem_limit);
+        }
+
+        if let Some(cpu_request) = self.cpu_request {
+            requests.insert("cpu".into(), cpu_request);
+        }
+
+        if let Some(mem_request) = self.mem_request {
+            requests.insert("memory".into(), mem_request);
+        }
+
+        // Insert all other resources not covered by the with_cpu_* and
+        // with_memory_* methods.
+        for (resource, (rr_type, quantity)) in self.other {
+            match rr_type {
+                ResourceRequirementsType::Limits => limits.insert(resource, quantity),
+                ResourceRequirementsType::Requests => requests.insert(resource, quantity),
+            };
+        }
+
+        // Only add limits/requests when there is actually stuff to add
+        let limits = if !limits.is_empty() {
+            Some(limits)
+        } else {
+            None
+        };
+
+        let requests = if !requests.is_empty() {
+            Some(requests)
+        } else {
+            None
+        };
```

Comment:
This pattern is also used [here](https://github.com/stackabletech/operator-rs/blob/main/src/commons/resources.rs#L302-L317). I just adjusted the ordering, otherwise leaving as is.